proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,500 @@
+
+Pinniped could accept either Fine-grained PATs or Classic PATs (or both), at the discretion of the Supervisor admin.
+PATs of each of these types have different prefixes and can be distinguished from each other by the prefix.
+Fine-grained PATs would need to be created with the permission to read org memberships. Classic PATs would need to
+be created to allow the `read:user` and `read:org` scopes.
+
+However, using PATs for this purpose would have some important implications:
+
+1. At first glance, it might appear that the Supervisor would need to hold the PAT in its session storage
+   (until the Supervisor session has expired) so it can validate the user again upon session refresh.
+   However, this PAT would be usable to grant access to Kubernetes
+   clusters to anyone who holds it, similar to an end-user password for CLI-based (non-interactive) OIDC or
+   LDAP/AD logins. The Supervisor does not store those OIDC/LDAP/AD passwords, and it should not store these
+   PATs either, if possible.
+2. PATs have no concept of intended audience, so there is no way for a user to consent that one particular PAT is
+   intended to control their access to Kubernetes clusters. A GitHub user can have many PATs for different purposes,
+   and we should not treat them as interchangeable.
+
+Solving the above points are key to designing non-interactive auth for GitHubIdentityProviders.
+
+##### How to Avoid Storing PATs in the Supervisor
+
+We could avoid storing PATs on the Supervisor by simply preventing token refreshes for any Supervisor sessions that
+were started using a GitHub PAT. There's no need to store the PAT for the purposes of refreshing the session if
+there is no concept of refreshing these sessions. By always returning an error on these refresh requests,
+the Pinniped CLI will automatically submit the PAT again (if stored in the user's env vars for the CLI process)
+to start a new session. There will be no observable user experience impact from always rejecting refreshes.
+
+By disabling refresh for these sessions, these users, when actively interacting with Kubernetes clusters, will
+create many more sessions per hour compared to users who can refresh their current sessions. Typical usage of
+Pinniped grants access to Kubernetes clusters for about five minutes before the next session refresh.
+This will cause two concerns:
+
+1. Each new session creates several Kubernetes Secrets for Supervisor session storage, so increasing the number of
+   sessions will increase the number of Secrets.
+2. Each new session will need to make several calls to the GitHub API, impacting the GitHub user's hourly rate limit.
+
+We can mitigate both effects using the techniques outlined in https://github.com/vmware-tanzu/pinniped/pull/1857.
+By making the access token lifetime slightly longer, the user will be able to use their clusters for about 10 minutes,
+rather than 5 minutes, reducing the number of new sessions that they need to start per hour.
```

Comment:
I asked a question about this on the POC: https://github.com/vmware-tanzu/pinniped/pull/1857/files#r1481728020 (linking for cross-reference value).